Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -617,6 +678,111 @@ private void ReceivePing(Ping ping)
+            ConcurrentBag<BoundPeer> history,
+            IEnumerable<BoundPeer> found,
+            Address target,
+            int depth,
+            TimeSpan? timeout,
+            CancellationToken cancellationToken,
+            Address searchAddress)
+        {
+            BoundPeer peerFound = null;
+            List<BoundPeer> peers = found.Where(
+                peer => !peer.Address.Equals(_address)).ToList();
+
+            if (peers.Count == 0)
+            {
+                _logger.Debug("No any neighbor received.");
+                return peerFound;
+            }
+
+            peers = Kademlia.SortByDistance(peers, target);
+
+            List<BoundPeer> closestNeighbors = _routing.Neighbors(target, _bucketSize).ToList();
+
+            Task[] awaitables = peers.Select(peer =>
+                PingAsync(peer, _requestTimeout, cancellationToken)
+            ).ToArray();
+            try
+            {
+                await Task.WhenAll(awaitables);
+            }
+            catch (AggregateException e)
+            {
+                if (e.InnerExceptions.All(ie => ie is TimeoutException) &&
+                    e.InnerExceptions.Count == awaitables.Length)
+                {
+                    throw new TimeoutException(
+                        $"All neighbors found do not respond in {_requestTimeout}."
+                    );
+                }
+
+                _logger.Error(
+                    e,
+                    "Some responses from neighbors found unexpectedly terminated: {Exception}",
+                    e
+                );
+            }
+
+            for (int i = 0; i < closestNeighbors.Count; i++)
+            {
+                if (string.CompareOrdinal(
+                        closestNeighbors[i].Address.ToHex(),
+                        searchAddress.ToHex()
+                   ) == 0)
+                {
+                    if (_routing.Contains(closestNeighbors[i]))
+                    {
+                        peerFound = closestNeighbors[i];
+                        return peerFound;
+                    }
+                }
+            }
+
+            var findNeighboursTasks = new List<Task<BoundPeer>>();
+            Peer closestKnown = closestNeighbors.Count == 0 ? null : closestNeighbors[0];
+            for (int i = 0; i < Kademlia.FindConcurrency && i < peers.Count; i++)
+            {
+                if (peerFound is null ||
+                    string.CompareOrdinal(
+                        Kademlia.CalculateDistance(peers[i].Address, target).ToHex(),
+                        Kademlia.CalculateDistance(closestKnown.Address, target).ToHex()
+                    ) < 1)
+                {
+                    findNeighboursTasks.Add(FindSpecificPeerAsync(
+                        history,
+                        target,
+                        peers[i],
+                        (depth == -1) ? depth : depth + 1,
```

Comment:
```suggestion
                        (depth == -1) ? depth : depth - 1,
```
Your logic in [KademliaProtocol.cs#L313](https://github.com/planetarium/libplanet/pull/580/files#diff-ae527d5afcdf41da4a57b594cab3cfefR313) expects this.